### PR TITLE
Add 'inlined from' annotations to locations in diagnostics about code that has been inlined

### PR DESCRIPTION
**Pull request by brmataptos**
### Description

Add 'inlined from' annotations to locations in diagnostics about code that has been inlined in move-compiler-v2

- Add an `Option<Box<Loc>>` field `inlined_from_loc` to struct `Loc` in move-model/src/model.rs, modify diagnostics to show it if present.
- Use it in move-compiler-v2/src/inliner.rs to augment code with information about where it was inlined.

### Test Plan
- Existing test cases private_call.move, private_call_2.move, and private_call_3.move show improved information in this case.
- Other test cases show lack of disruption.

### Fixes
Bug #10848

### CODE REVIEW

File: third_party/move/move-model/src/model.rs [hunk 1]
```diff
@@ -111,7 +112,31 @@ impl AsRef<Loc> for Loc {
...
+            inlined_from_loc: Some(Box::new(match &self.inlined_from_loc {
+                None => inlined_from.clone(),
+                Some(locbox) => (*locbox.clone()).inlined_from(inlined_from),
+            })),
+        }
+    }
+
+    fn inline_if_needed(&self, loc: Loc) -> Loc {
+        if let Some(locbox) = &self.inlined_from_loc {
+            let source_loc = &**locbox;
```
Comment (vineethk) [inline third_party/move/move-model/src/model.rs:135]:
```suggestion
            let source_loc = locbox.as_ref();
```

File: third_party/move/move-model/src/model.rs [hunk 2]
```diff
@@ -111,7 +112,31 @@ impl AsRef<Loc> for Loc {
...
+            file_id: self.file_id,
+            span: self.span,
+            inlined_from_loc: Some(Box::new(match &self.inlined_from_loc {
+                None => inlined_from.clone(),
+                Some(locbox) => (*locbox.clone()).inlined_from(inlined_from),
+            })),
+        }
+    }
+
+    fn inline_if_needed(&self, loc: Loc) -> Loc {
```
Comment (vineethk) [inline third_party/move/move-model/src/model.rs:133]: Comments for these methods would be helpful.

File: third_party/move/move-model/src/model.rs [hunk 3]
```diff
@@ -797,21 +824,36 @@ impl GlobalEnv {
         self.diag_with_notes(Severity::Error, loc, msg, notes)
     }
 
+    fn add_inlined_from_labels(labels: &mut Vec<Label<FileId>>, inlined_from: &Option<Box<Loc>>) {
```
Comment (vineethk) [inline third_party/move/move-model/src/model.rs:827]: Method doc, please.

File: third_party/move/move-model/src/model.rs [hunk 4]
```diff
@@ -797,21 +824,36 @@ impl GlobalEnv {
...
+        let mut inlined_from = inlined_from;
+        while let Some(boxed_loc) = inlined_from {
+            let loc = &**boxed_loc;
```
Comment (vineethk) [inline third_party/move/move-model/src/model.rs:830]:
```suggestion
            let loc = boxed_loc.as_ref();
```

File: third_party/move/move-model/src/model.rs [hunk 5]
```diff
@@ -876,6 +919,7 @@ impl GlobalEnv {
         Loc {
```
Comment (vineethk) [inline third_party/move/move-model/src/model.rs:919]: Is there a reason not to use `Loc::new` here?

File: third_party/move/move-compiler-v2/tests/checking/inlining/private_call_3.exp [hunk 6]
```diff
@@ -175,49 +175,199 @@ module 0x42::n {
...
-  │         ^^^^^
+   ┌─ tests/checking/inlining/private_call_3.move:5:9
+   │
+ 5 │         bar()
+   │         ^^^^^
+   ·
+67 │         m::foo();
+   │         -------- in a call inlined at this callsite
+   ·
+98 │         assert!(o_nonfriend::foo() == 42, 1);
```
Comment (vineethk) [inline third_party/move/move-compiler-v2/tests/checking/inlining/private_call_3.exp:186]: Nice!

File: third_party/move/move-compiler-v2/tests/checking/inlining/lambda_param.exp [hunk 7]
```diff
@@ -11,3 +11,6 @@ error: Invalid call target: currently indirect call must be a parameter to an in
   │
```
Comment (vineethk) [inline third_party/move/move-compiler-v2/tests/checking/inlining/lambda_param.exp:11]: This seems like a duplicate error: would the two errors here (at line 7 above and this one) always be coupled (i.e., always appear in pairs)? If so, we could remove reporting the error with "indirect call" and only keep the one above?

File: third_party/move/move-model/src/ast.rs [hunk 8]
```diff
@@ -1007,6 +1007,30 @@ impl ExpData {
         }
     }
 
+    /// A function which can be used by a `node_rewriter` argument to `ExpData::rewrite_node_id` to
+    /// update node location (`Loc`), in addition to instantiating types.
```
Comment (vineethk) [inline third_party/move/move-model/src/ast.rs:1011]: Maybe also add some guidance on when to use which method (this one vs. the one above)?

**Review by vineethk — COMMENTED** [review]
Looking good so far, I have a few minor comments.

File: third_party/move/move-model/src/model.rs (hunk 5 above)
Comment (brmataptos) [reply third_party/move/move-model/src/model.rs:919]: It was already written that way.  But I've updated it.

File: third_party/move/move-compiler-v2/tests/checking/inlining/lambda_param.exp (hunk 7 above)
Comment (brmataptos) [reply third_party/move/move-compiler-v2/tests/checking/inlining/lambda_param.exp:11]: Ok, done.

File: third_party/move/move-model/src/ast.rs (hunk 8 above)
Comment (brmataptos) [reply third_party/move/move-model/src/ast.rs:1011]: Added something.

**Review by wrwg — APPROVED** [review]
Nice error messages.

Bit painful to see `Loc` growing as this thing is used and copied all over the place, but probably not feasible way around it.

**Review by rahxephon89 — APPROVED** [review]

**Pull request merged**